torch/_functorch/_aot_autograd/streams.py

```diff
@@ -67,6 +108,147 @@ def insert_wait_event_before_node(graph: Graph, node: Node, event_ind: int) -> N
+    graph: Graph,
+    stream_index: Optional[int],
+) -> IndexedDict[Node, float]:
+    if stream_index not in stream_to_timeline:
+        stream_to_timeline[stream_index] = IndexedDict()
+        total_time = 0.0
+        for node in graph.nodes:
+            # mlazos: not sure if we should include forward here too but don't think it matters
+            if is_bwd_node(node) and get_stream(node) == stream_index:
+                total_time += get_roofline_estimate(node)
+                stream_to_timeline[stream_index][node] = (
+                    total_time  # NB: total time includes the node's runtime
+                )
+
+    return stream_to_timeline[stream_index]
+
+
+# NB: we start all estimates at 0, estimating the total runtime of each stream with timestamps at each node
+# we then try and use these timestamps to estimate when to deallocate tensors used in side streams
+# See https://docs.pytorch.org/docs/stable/generated/torch.Tensor.record_stream.html#torch.Tensor.record_stream
+# for details on the problem being addressed. Rather than using the automatic memory management approach of record_stream
+# we attempt to find the point which to deallocate based on the estimated timestamps.
+def handle_synced_deallocation(
+    graph: Graph,
+    stream_to_exec_trace: dict[Optional[int], IndexedDict[Node, float]],
+    node: Node,
+    last_usage: Node,
+) -> None:
+    assert is_bwd_node(node), (
+        "synced allocations should only be handled on backward nodes"
+    )
+    assert is_bwd_node(last_usage), (
+        "synced allocations should only be handled on backward nodes"
+    )
+    allocating_stream = get_stream(node)
+    side_stream = get_stream(last_usage)
+    assert allocating_stream != side_stream, (
+        "allocating and side stream should be different for synced deallocations"
+    )
+    if not torch.cuda.is_available():
+        # fallback to record_stream in this case
+        with graph.inserting_after(node):
+            graph.call_function(
+                torch.ops.streams.record_stream.default,
+                (
+                    node,
+                    get_stream_or_current_stream(last_usage),
+                ),
+                {},
+            )
+        node.meta["partitioner_tag"] = "must_be_in_backward"
+
+    allocating_stream_trace = populate_stream_timeline(
+        stream_to_exec_trace, graph, allocating_stream
+    )
+    side_stream_trace = populate_stream_timeline(
+        stream_to_exec_trace, graph, side_stream
+    )
+
+    alloc_ptr = node
+    target_side_stream_time = side_stream_trace[last_usage]
+    # linear search from first usage of tensor to a point in time after the side stream has finished
+    while alloc_ptr is not None:
+        alloc_time = allocating_stream_trace[alloc_ptr]
+
+        if alloc_time >= target_side_stream_time:
+            break
+        elif alloc_time < target_side_stream_time:
+            next_ptr = allocating_stream_trace.next_key(alloc_ptr)
```

Comment:
Instead of adding a new data structure - why don't we have a 

List[Tuple[fx.Node], timestamp]], and then a dict from node-> index ? the extra deallocations we're adding shouldn't affect this list right ?

Reply:
I had this originally actually - I think the datastructure version was more readable and organized (easier than passing both the list and dict around)